File: truevfs-ext/truevfs-ext-logging/src/main/scala/net/java/truevfs/ext/logging/LogMediator.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.logging

import java.io._
import java.nio.channels._
import net.java.truevfs.comp.inst._
import net.java.truevfs.kernel.spec._
import net.java.truecommons.cio._

/**
 * @author Christian Schlichtherle
 */
private class LogMediator extends Mediator[LogMediator] with Immutable {

  override def instrument(subject: FsManager): FsManager =
    new InstrumentingManager[LogMediator](this, subject)

  override def instrument(subject: IoBufferPool): IoBufferPool =
    new InstrumentingBufferPool[LogMediator](this, subject)

  override def instrument(context: InstrumentingManager[LogMediator], subject: FsCompositeDriver): FsCompositeDriver =
    new InstrumentingCompositeDriver[LogMediator](this, subject)

  override def instrument(context: InstrumentingBufferPool[LogMediator], subject: IoBuffer): IoBuffer =
    new LogBuffer(this, subject)

  override def instrument(context: InstrumentingCompositeDriver[LogMediator], subject: FsController): FsController =
    new InstrumentingController[LogMediator](this, subject)

  override def instrument[E <: Entry](context: InstrumentingController[LogMediator], subject: InputSocket[E]): InputSocket[E] =
    new InstrumentingInputSocket[LogMediator, E](this, subject)

  override def instrument[E <: Entry](context: InstrumentingController[LogMediator], subject: OutputSocket[E]): OutputSocket[E] =
    new InstrumentingOutputSocket[LogMediator, E](this, subject)

  override def instrument[B <: IoBuffer](context: InstrumentingBuffer[LogMediator], subject: InputSocket[B]): InputSocket[B] =
    new InstrumentingInputSocket[LogMediator, B](this, subject)

  override def instrument[B <: IoBuffer](context: InstrumentingBuffer[LogMediator], subject: OutputSocket[B]): OutputSocket[B] =
    new InstrumentingOutputSocket[LogMediator, B](this, subject)

  override def instrument[E <: Entry](context: InstrumentingInputSocket[LogMediator, E], subject: InputStream): InputStream =
    new LogInputStream(context, subject)

  override def instrument[E <: Entry](context: InstrumentingInputSocket[LogMediator, E], subject: SeekableByteChannel): SeekableByteChannel =
    new LogInputChannel(context, subject)

  override def instrument[E <: Entry](context: InstrumentingOutputSocket[LogMediator, E], subject: OutputStream): OutputStream =
    new LogOutputStream(context, subject)

  override def instrument[E <: Entry](context: InstrumentingOutputSocket[LogMediator, E], subject: SeekableByteChannel): SeekableByteChannel =
    new LogOutputChannel(context, subject)
}

private object LogMediator extends LogMediator

File: src/main/java/global/namespace/truevfs/ext/logging/LogMediator.java
/*
 * Copyright © 2005 - 2021 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package global.namespace.truevfs.ext.logging;

import global.namespace.truevfs.comp.cio.*;
import global.namespace.truevfs.comp.inst.*;
import global.namespace.truevfs.kernel.api.FsCompositeDriver;
import global.namespace.truevfs.kernel.api.FsController;
import global.namespace.truevfs.kernel.api.FsManager;

import java.io.InputStream;
import java.io.OutputStream;
import java.nio.channels.SeekableByteChannel;

final class LogMediator extends Mediator<LogMediator> {

    static final Mediator<?> SINGLETON = new LogMediator();

    private LogMediator() {
    }

    @Override
    public FsManager instrument(FsManager subject) {
        return new InstrumentingManager<>(this, subject);
    }

    @Override
    public IoBufferPool instrument(IoBufferPool subject) {
        return new InstrumentingBufferPool<>(this, subject);
    }

    @Override
    public FsCompositeDriver instrument(InstrumentingManager<LogMediator> context, FsCompositeDriver subject) {
        return new InstrumentingCompositeDriver<>(this, subject);
    }

    @Override
    public IoBuffer instrument(InstrumentingBufferPool<LogMediator> context, IoBuffer subject) {
        return new LogBuffer(this, subject);
    }

    @Override
    public FsController instrument(InstrumentingCompositeDriver<LogMediator> context, FsController subject) {
        return new InstrumentingController<>(this, subject);
    }

    @Override
    public <E extends Entry> InputSocket<E> instrument(InstrumentingController<LogMediator> context, InputSocket<E> subject) {
        return new InstrumentingInputSocket<>(this, subject);
    }

    @Override
    public <E extends Entry> OutputSocket<E> instrument(InstrumentingController<LogMediator> context, OutputSocket<E> subject) {
        return new InstrumentingOutputSocket<>(this, subject);
    }

    @Override
    public <B extends IoBuffer> InputSocket<B> instrument(InstrumentingBuffer<LogMediator> context, InputSocket<B> subject) {
     return   new InstrumentingInputSocket<>(this, subject);
    }

    @Override
    public <B extends IoBuffer> OutputSocket<B> instrument(InstrumentingBuffer<LogMediator> context, OutputSocket<B> subject) {
        return new InstrumentingOutputSocket<>(this, subject);
    }

    @Override
    public <E extends Entry> InputStream instrument(InstrumentingInputSocket<LogMediator,E> context, InputStream subject) {
        return new LogInputStream(context, subject);
    }

    @Override
    public <E extends Entry> SeekableByteChannel instrument(InstrumentingInputSocket<LogMediator,E> context, SeekableByteChannel subject) {
        return new LogInputChannel(context, subject);
    }

    @Override
    public <E extends Entry> OutputStream instrument(InstrumentingOutputSocket<LogMediator,E> context, OutputStream subject) {
        return new LogOutputStream(context, subject);
    }

    @Override
    public <E extends Entry> SeekableByteChannel instrument(InstrumentingOutputSocket<LogMediator,E> context, SeekableByteChannel subject) {
        return new LogOutputChannel(context, subject);
    }
}
